### PR TITLE
Added current_auth. Fixes #154

### LINKED ISSUE #154: Switch from g.user to current_user
[Coaster currently expects a `g.user` entity in `load_models`](https://github.com/hasgeek/coaster/blob/9a7f456306323ea503d936a5e0526ce544edffe3/coaster/views.py#L267-L274). If the implementation for hasgeek/baseframe#80 (CRUD views) is moved to Coaster, that will add a second reference to `g.user`.

However, hasgeek/flask-lastuser#24 proposes replacing `g.user` with a `current_user` proxy, which must be explicitly imported. This is a problem because Flask-Lastuser depends on Coaster. Coaster can't import from Flask-Lastuser.

We can get around this cyclic dependency by making a choice:

1. Stick with `g.user`. The problems it causes will remain, such as the possibility of an app's `before_request` handler attempting to access it before it is set.

2. Coaster should add a dependency on Flask-Login and import `current_user` from there. Flask-Lastuser will need to plugin to Flask-Login and replace the user loading mechanism there.

3. Coaster should provide its own `current_user` proxy, but allow the backing implementation to be replaced (process-wide). Importing Flask-Lastuser into the same process will automatically apply this replacement.

### PR DESCRIPTION
**Pull request by jace**
This PR breaks compatibility between Flask-Lastuser and load_models/docflow (the two existing uses of `g.user`). Flask-Lastuser needs to be updated to be compatible with Coaster’s API before this PR can be merged.

### CODE REVIEW

**Comment by jace** [comment]
I'm ambivalent between providing access as `current_user.self` or `current_user.user`. Opinions?

**Comment by jace** [comment]
I propose renaming to `current_user.user` as this framework is very likely to grow to include other principals, anchors and tokens.

**Comment by jace** [comment]
On the other hand, since it offers transparent access to the user object's attributes, reserving attributes that are very likely to conflict with attributes in the user object is not a good idea. `self` is a very unlikely attribute to occur in user objects.

For example, when a client accesses a resource provider on behalf of a user, we may want the `current_user` to represent both the user object and the calling client. These would be logically available as `current_user.user` and `current_user.client`. However, at least in Lastuser, we already have a `user.client` attribute that indicates the client that created the user account—a fairly distinct concept.

The proxy needs to exercise care in the names of its direct attributes.

**Comment by jace** [comment]
Flask-Lastuser is now partially compatible with `current_user`. This is good enough to start replacing all uses of `g.user` with `current_user` in client apps. (Full compatibility requires providing `current_app.login_manager._load_user()`.)

**Comment by jace** [comment]
hasgeek/lastuser#153 describes a "sudo" framework where an admin can switch into a user's account. We will conceivably want both the user and the admin's account to be exposed via `current_user`. If `.self` refers to the user account, should the admin account be `.sudo`?

File: coaster/logger.py [hunk 1]
```diff
@@ -109,6 +110,10 @@ def formatException(self, ei):
             print("App context:", file=sio)
             pprint_with_indent(vars(g), sio)
 
+        print('\n----------\n', file=sio)
+        print("Current user:", file=sio)
+        pprint_with_indent(vars(current_user), sio)
```
Comment (jace) [inline coaster/logger.py]: This currently renders as `{'self': <User MFLi1lfRR2eTpqRj1fXe8g (jace) "Kiran Jonnalagadda">}`. If we use `vars(current_user.self)`, we get too much unnecessary detail on the user object. Rendering `current_user` alone will also capture other variables within it as it grows in scope.

**Comment by jace** [comment]
If we add an anchor framework in Coaster (briefly discussed in #156, full ticket pending), then we may want to collate inbound anchors, perhaps as `current_user.anchors`. However, since anchors aren't properly defined yet, they should be outside the scope of this PR, save for this note on having a name reserved for them.

**Comment by jace** [comment]
The list of attributes we imagine will need to be available from `current_user` (or `current_auth`) so far is:

1. `is_authenticated` (bool)
2. `is_anonymous` (bool)
3. `permissions` (list of permissions available to the current user in the current request)
4. `anchors` (list of inbound anchors associated with this request)
5. `self` or `user` (object, as are all of the following)
6. `client` (for client calling on behalf of a user; sometimes a client will be present but no user)
7. `sudo` (for the admin user currently operating another user's account)

**Comment by jace** [comment]
Given all these possible attributes existing only some of the time, and the pain of doing `hasattr(current_user, attr_name)`, I propose `CurrentAuth` should implement a `__getattr__` that simply returns `None`. It'll only be called when the attribute does not exist.

Callers still have to do a conditional check, which is now `if current_auth.attr_name`.

File: tests/test_auth.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    def test_current_auth_with_user_unloaded(self):
+        user = User(username='foo', fullname='Mr Foo')
+        self.login_manager.set_user_for_testing(user)
+
+        self.assertFalse(current_auth.is_anonymous)
+        self.assertTrue(current_auth.is_authenticated)
+        self.assertIsNotNone(current_auth.user)
+        self.assertEqual(current_auth.user, user)
+
+        # Additional auth details (username only in thi test) exposed by the login manager
```
Comment (shreyas-satish) [inline tests/test_auth.py]: Typo, think you meant `this`.

**Comment by jace** [comment]
I've made a late call to expose `current_auth` to templates from Coaster instead of Flask-Lastuser as it no longer conflicts with `current_user` from any other source, and it seems like a more logical place.

**Pull request merged**